### PR TITLE
AWS DDB SDK client support for remote data store

### PR DESCRIPTION
**Pull request by arjunkumargiri**
### Description
Add AWS DDB implementation of SDK client. This change would allow plugin metadata to be written to AWS DDB tables. Users need to set environment variable `REMOTE_METADATA_TYPE` as `AWSDynamoDB` to use DDB as remote plugin metadata store.
 
 
### Check List
- [ ] New functionality includes testing.
  - [x ] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+     */
+    @Override
+    public CompletionStage<PutDataObjectResponse> putDataObjectAsync(PutDataObjectRequest request, Executor executor) {
+        final String id = request.id() != null ? request.id() : UUID.randomUUID().toString();
+        final String tenantId = request.tenantId() != null ? request.tenantId() : DEFAULT_TENANT;
+        final String tableName = getTableName(request.index());
+        return CompletableFuture.supplyAsync(() -> AccessController.doPrivileged((PrivilegedAction<PutDataObjectResponse>) () -> {
+            try (XContentBuilder sourceBuilder = XContentFactory.jsonBuilder()) {
+                XContentBuilder builder = request.dataObject().toXContent(sourceBuilder, ToXContent.EMPTY_PARAMS);
+                String source = builder.toString();
```
Comment (dbwiddis) [inline plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:75]: I found out later there's a util for this
```suggestion
            try {                
                String source = Strings.toString(MediaTypeRegistry.JSON, request.dataObject());
```

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+                        Map.entry(SOURCE, AttributeValue.builder().s(source).build())
+                    );
+                final PutItemRequest putItemRequest = PutItemRequest.builder().tableName(tableName).item(item).build();
+
+                dynamoDbClient.putItem(putItemRequest);
+                return new PutDataObjectResponse.Builder().id(id).created(true).build();
+            } catch (Exception e) {
+                log.error("Exception while inserting data into DDB: " + e.getMessage(), e);
+                throw new OpenSearchException(e);
+            }
```
Comment (dbwiddis) [inline plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:90]: Probably want better exception handling than this, see patterns in https://github.com/opensearch-project/ml-commons/pull/2507

Better yet, probably want to merge that and rebase this PR CC: @dhrubo-os

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+                        Map.entry(RANGE_KEY, AttributeValue.builder().s(request.id()).build())
+                    )
+            )
+            .build();
+        return CompletableFuture.supplyAsync(() -> AccessController.doPrivileged((PrivilegedAction<DeleteDataObjectResponse>) () -> {
+            dynamoDbClient.deleteItem(deleteItemRequest);
+            return new DeleteDataObjectResponse.Builder().id(request.id()).deleted(true).build();
+        }), executor);
+    }
+
```
Comment (dbwiddis) [inline plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:153]: Can add Update once https://github.com/opensearch-project/ml-commons/pull/2520 is merged.

Probably sould merge that and rebase this PR. CC: @dhrubo-os

**Review by dbwiddis — COMMENTED** [review]
LGTM so far.  Need to catch up with PRs awaiting review/merging.

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java (hunk 3 above)
Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:153]: Do we need to merge both `2507` & `2520`? Seems like `2520` includes `2507` as well?

File: plugin/src/main/java/org/opensearch/ml/sdkclient/SdkClientModule.java [hunk 4]
```diff
@@ -16,48 +16,89 @@
...
+        this(System.getenv(REMOTE_METADATA_TYPE), System.getenv(REMOTE_METADATA_ENDPOINT), System.getenv(REGION));
     }
 
     /**
      * Instantiate this module specifying the endpoint and region. Package private for testing.
      * @param remoteMetadataEndpoint The remote endpoint
      * @param region The region
      */
-    SdkClientModule(String remoteMetadataEndpoint, String region) {
+    SdkClientModule(String remoteStoreType, String remoteMetadataEndpoint, String region) {
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/sdkclient/SdkClientModule.java:60]: should we keep similarity with `REMOTE_METADATA_TYPE` like `remoteMetadataType`?

File: common/src/main/java/org/opensearch/sdk/DeleteDataObjectRequest.java [hunk 5]
```diff
@@ -41,12 +44,17 @@ public String id() {
         return this.id;
     }
 
+    public String tenantId() {
```
Comment (dbwiddis) [inline common/src/main/java/org/opensearch/sdk/DeleteDataObjectRequest.java:47]: Needs a javadoc.

Won't repeat this but almost everything added here has no javadocs.

File: common/src/main/java/org/opensearch/sdk/DeleteDataObjectRequest.java [hunk 6]
```diff
@@ -73,12 +81,17 @@ public Builder id(String id) {
             return this;
         }
 
+        public Builder tenantId(String tenantId) {
+            this.tenantId = tenantId;
+            return this;
+        }
```
Comment (dbwiddis) [inline common/src/main/java/org/opensearch/sdk/DeleteDataObjectRequest.java:87]: I like how we're demonstrating how easy it is to add to these request objects to increase functionality.

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    /**
+     * DDB implementation to write data objects to DDB table. Tenant ID will be used as hash key and document ID will
+     * be used as range key. If tenant ID is not defined a default tenant ID will be used. If document ID is not defined
+     * a random UUID will be generated. Data object will be written as a nested DDB attribute.
+     *
+     */
+    @Override
+    public CompletionStage<PutDataObjectResponse> putDataObjectAsync(PutDataObjectRequest request, Executor executor) {
+        final String id = request.id() != null ? request.id() : UUID.randomUUID().toString();
+        final String tenantId = request.tenantId() != null ? request.tenantId() : DEFAULT_TENANT;
```
Comment (dbwiddis) [inline plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:70]: curious if `DEFAULT_TENANT` should be in the request object to start with.

File: plugin/src/main/java/org/opensearch/ml/sdkclient/SdkClientModule.java [hunk 8]
```diff
@@ -16,48 +16,89 @@
...
+    private final String remoteStoreType;
     private final String remoteMetadataEndpoint;
     private final String region; // not using with RestClient
 
     /**
      * Instantiate this module using environment variables
      */
     public SdkClientModule() {
-        this(System.getenv(REMOTE_METADATA_ENDPOINT), System.getenv(REGION));
+        this(System.getenv(REMOTE_METADATA_TYPE), System.getenv(REMOTE_METADATA_ENDPOINT), System.getenv(REGION));
```
Comment (dbwiddis) [inline plugin/src/main/java/org/opensearch/ml/sdkclient/SdkClientModule.java:52]: Consider changing these to OpenSearch properties.

Comment (arjunkumargiri) [reply plugin/src/main/java/org/opensearch/ml/sdkclient/SdkClientModule.java:52]: Agreed. Let's have a separate task to migrate from environment property to opensearch property.

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java (hunk 7 above)
Comment (arjunkumargiri) [reply plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:73]: `DEFAULT_TENANT` will not be required for other data store implementations. Given that it is going to be specific to DDB store I would prefer it to be part of `DDBOpenSearchClient.java`

**Comment by dhrubo-os** [comment]
@arjunkumargiri somehow your Pull request has all of our commits.

**Comment by arjunkumargiri** [comment]
> @arjunkumargiri somehow your Pull request has all of our commits.

Checking. I did a rebase to pull in the latest changes but not sure why those are included in my PR.

File: plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java (hunk 3 above)
Comment (arjunkumargiri) [reply plugin/src/main/java/org/opensearch/ml/sdkclient/DDBOpenSearchClient.java:155]: Will add support for update and search as part of new PR.

**Review by dhrubo-os — APPROVED** [review]

**Pull request merged**